File: mobile/lib/models/match.dart
/// Data model for a football match.

class MatchModel {
  final int id;
  final int fixtureId;
  final String leagueName;
  final int leagueId;
  final String leagueLogo;
  final String country;
  final String homeTeam;
  final String awayTeam;
  final String homeLogo;
  final String awayLogo;
  final String status;
  final int elapsed;
  final int scoreHome;
  final int scoreAway;
  final String kickOff;
  final double oddHome;
  final double oddDraw;
  final double oddAway;
  final int tournamentType;
  final bool isLive;

  MatchModel({
    required this.id,
    required this.fixtureId,
    this.leagueName = '',
    this.leagueId = 0,
    this.leagueLogo = '',
    this.country = '',
    this.homeTeam = '',
    this.awayTeam = '',
    this.homeLogo = '',
    this.awayLogo = '',
    this.status = 'NS',
    this.elapsed = 0,
    this.scoreHome = 0,
    this.scoreAway = 0,
    this.kickOff = '',
    this.oddHome = 0,
    this.oddDraw = 0,
    this.oddAway = 0,
    this.tournamentType = 0,
    this.isLive = false,
  });

  factory MatchModel.fromJson(Map<String, dynamic> json) {
    return MatchModel(
      id: json['id'] ?? 0,
      fixtureId: json['fixture_id'] ?? 0,
      leagueName: json['league_name'] ?? '',
      leagueId: json['league_id'] ?? 0,
      leagueLogo: json['league_logo'] ?? '',
      country: json['country'] ?? '',
      homeTeam: json['home_team'] ?? '',
      awayTeam: json['away_team'] ?? '',
      homeLogo: json['home_logo'] ?? '',
      awayLogo: json['away_logo'] ?? '',
      status: json['status'] ?? 'NS',
      elapsed: json['elapsed'] ?? 0,
      scoreHome: json['score_home'] ?? 0,
      scoreAway: json['score_away'] ?? 0,
      kickOff: json['kick_off'] ?? '',
      oddHome: (json['odd_home'] ?? 0).toDouble(),
      oddDraw: (json['odd_draw'] ?? 0).toDouble(),
      oddAway: (json['odd_away'] ?? 0).toDouble(),
      tournamentType: json['tournament_type'] ?? 0,
      isLive: json['is_live'] ?? false,
    );
  }

  bool get isNotStarted => status == 'NS' || status == 'TBD';
  bool get isFinished => status == 'FT' || status == 'AET' || status == 'PEN';
}

/// A group of matches under one league.
class LeagueGroup {
  final String leagueName;
  final String leagueLogo;
  final String country;
  final List<MatchModel> matches;

  LeagueGroup({
    required this.leagueName,
    this.leagueLogo = '',
    this.country = '',
    required this.matches,
  });

  factory LeagueGroup.fromJson(Map<String, dynamic> json) {
    return LeagueGroup(
      leagueName: json['league_name'] ?? '',
      leagueLogo: json['league_logo'] ?? '',
      country: json['country'] ?? '',
      matches: (json['matches'] as List? ?? [])
          .map((m) => MatchModel.fromJson(m))
          .toList(),
    );
  }
}

File: backend/app/models/match.py
import datetime
from sqlalchemy import String, Integer, Float, DateTime, Boolean, JSON
from sqlalchemy.orm import Mapped, mapped_column
from app.models.database import Base


class Match(Base):
    __tablename__ = "matches"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    fixture_id: Mapped[int] = mapped_column(Integer, unique=True, index=True)

    # League info
    league_name: Mapped[str] = mapped_column(String(200), default="")
    league_id: Mapped[int] = mapped_column(Integer, default=0)
    league_logo: Mapped[str] = mapped_column(String(500), default="")
    country: Mapped[str] = mapped_column(String(100), default="")

    # Teams
    home_team: Mapped[str] = mapped_column(String(200), default="")
    away_team: Mapped[str] = mapped_column(String(200), default="")
    home_logo: Mapped[str] = mapped_column(String(500), default="")
    away_logo: Mapped[str] = mapped_column(String(500), default="")

    # Status: NS (not started), 1H, HT, 2H, FT, etc.
    status: Mapped[str] = mapped_column(String(10), default="NS")
    elapsed: Mapped[int] = mapped_column(Integer, default=0)

    # Score
    score_home: Mapped[int] = mapped_column(Integer, default=0)
    score_away: Mapped[int] = mapped_column(Integer, default=0)

    # Kick-off time (UTC)
    kick_off: Mapped[datetime.datetime] = mapped_column(
        DateTime, default=datetime.datetime.utcnow
    )

    # Odds (pre-match)
    odd_home: Mapped[float] = mapped_column(Float, default=0.0)
    odd_draw: Mapped[float] = mapped_column(Float, default=0.0)
    odd_away: Mapped[float] = mapped_column(Float, default=0.0)

    # Tournament type: 0 = league, 1 = cup
    tournament_type: Mapped[int] = mapped_column(Integer, default=0)

    # Live stats (updated during match)
    shots_home: Mapped[int] = mapped_column(Integer, default=0)
    shots_away: Mapped[int] = mapped_column(Integer, default=0)
    corners_home: Mapped[int] = mapped_column(Integer, default=0)
    corners_away: Mapped[int] = mapped_column(Integer, default=0)
    red_cards_home: Mapped[int] = mapped_column(Integer, default=0)
    red_cards_away: Mapped[int] = mapped_column(Integer, default=0)

    # Is it currently live?
    is_live: Mapped[bool] = mapped_column(Boolean, default=False)

    # Extra metrics (Possession, Fouls, Yellow Cards, Offsides, etc.)
    extra_stats: Mapped[dict] = mapped_column(JSON, default=dict)

    created_at: Mapped[datetime.datetime] = mapped_column(
        DateTime, default=datetime.datetime.utcnow
    )

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "fixture_id": self.fixture_id,
            "league_name": self.league_name,
            "league_id": self.league_id,
            "league_logo": self.league_logo,
            "country": self.country,
            "home_team": self.home_team,
            "away_team": self.away_team,
            "home_logo": self.home_logo,
            "away_logo": self.away_logo,
            "status": self.status,
            "elapsed": self.elapsed,
            "score_home": self.score_home,
            "score_away": self.score_away,
            # Stamp the stored (naive UTC) kick-off with an explicit UTC marker so
            # clients parse it as UTC and convert to local time correctly. Without
            # the marker the app reads UTC as local and shows times 3h early.
            "kick_off": self.kick_off.replace(tzinfo=datetime.timezone.utc).isoformat() if self.kick_off else None,
            "odd_home": self.odd_home,
            "odd_draw": self.odd_draw,
            "odd_away": self.odd_away,
            "tournament_type": self.tournament_type,
            "shots_home": self.shots_home,
            "shots_away": self.shots_away,
            "corners_home": self.corners_home,
            "corners_away": self.corners_away,
            "red_cards_home": self.red_cards_home,
            "red_cards_away": self.red_cards_away,
            "extra_stats": self.extra_stats or {},
            "is_live": self.is_live,
        }
